👌 IMPROVE: enable custom builder passthrough

mmcky opened the pull request:
This PR enables a `builder` to pass through `jupyter-book` which can be useful when using sphinx extensions that introduce a new build output (that we don't necessarily want to fully support as a builder directly). This is only useful for `advanced` users that have a good working knowledge of `sphinx` and wishing to use `jupyter-book` interface to build new outputs. 

**Example:**

```bash
jb build lectures --builder=custom --custom-name=<builder-name>
```

**Background:**

I have been working on [sphinx-tojupyter](https://github.com/QuantEcon/sphinx-tojupyter) which is **only intended to be a short to medium term solution** to build notebooks that only contain basic `markdown` instead of the full `myst` syntax. Once `myst` is supported by `jupyter` I will deprecate this package.

This is useful for publishing `download` notebooks to support a project that may wish to distribute `notebooks`. I won't sink a lot of development time into this tool -- it is basically a migration of `writer/translators` from the old QuantEcon project `sphinxcontrib-tojupyter`. It adds a `jupyter` builder for building notebooks.

mmcky commented:
@choldgraf would you have a moment to review this and let me know if this can be merged? I'd like to use to to build via `jupyter-book` an alternative builder but not one we want to officially support. Just passthrough.

choldgraf commented:
@mmcky I wonder if it would be easier to not add an extra config value, and instead to do something like support:

```
jupyter-book build mybook/ --builder=someotherbuilder --unknown-builder
```

And the `--unknown-builder` flag would remove the check for whether it was a "known" builder, and then it would follow the rest of the logic that you've included in this PR. Does that seem like it would be conceptually easier to wrap one's head around? (we could also call it `--force-builder` or something like that)

mmcky commented:
@choldgraf I like your thinking on this. I'll rework this now. I might call it `--custom-builder`?

```
jupyter-book build mybook/ --builder=someotherbuilder --custom-builder
```

and that flag would enable it to passthrough

mmcky commented:
@choldgraf what do you think about using `--external-builder` -- that seems most descriptive to me.

mmcky commented:
I used the previous structure as I wasn't sure how we could update `BUILDER_OPTS` on the fly with a custom name via `click.Options` -- will have a think.

mmcky commented:
@choldgraf I update this to use ``--external-builder`` but that can be easily changed.

I have changed `--builder` click option to accept any string rather than `click.Options` and added a bit of logic to check the option ourselves in `build`. This enables checking of `external_builder` and updating the `BUILDER_OPTS` dictionary prior to parsing the `--builder`

choldgraf commented:
Hmmm - I think the one downside there is that now the options are printed when people do `jupyter-book build --help`.

Given this limitation, I'm gonna be annoying and suggest maybe we go back to your original proposal, but slightly modified? What about something like:

```
jb build --builder=custom --custom-builder=<builder-name>
```

?

mmcky commented:
thanks @choldgraf -- these comments are great. This is a nice syntax -- I have made this change.

mmcky commented:
@choldgraf not entirely sure how to test this without installing `sphinx-tojupyter` and building a sample project. I am not entirely sure it's worth it given how few will likely be using this feature. Thoughts? `codecov` is the only failure.

choldgraf commented:
hmmmm - perhaps we could just add a test site that:

1. Defines a custom builder that is just a copy of the HTMLBuilder
2. Registers that builder as "mycustombuilder"
3. The test itself builds the site with `--builder customer --custom-builder mycustombuilder`
4. There is a check on `sphinx.app` to make sure the builder is the one we expect

That could be a lightweight way to try the functionality? I *think* we have access to `sphinx.app` in the tests

mmcky commented:
thanks @choldgraf - nice idea setting up an alias for the html builder. 
I guess the builder can now be registered through `local_extensions` now 🎉

mmcky commented on tests/test_build.py:
```diff
@@ -209,3 +209,19 @@ def test_execution_timeout(pages, build_resources, cli):
+    books, _ = build_resources
+    config = books.joinpath("config_custombuilder")
+    result = cli.invoke(
+        commands.build(builder="custom", custom_builder="mycustombuilder"),
```
@choldgraf @chrisjsewell while writing a test for this PR (adding custom builder passthrough) I am not sure where to specify the `jupyter-book` command line options. The test seems to fail when adding them as keyword arguments to `commands.build`.

mmcky replied:
Ah - I think I have them in the right place now

mmcky commented:
@chrisjsewell @choldgraf I have added a test to this that I think is now properly configured but it is not building with an error message that indicates it isn't picking up the `local_extensions` in the the `_config.yml` file. When I builder `config_custombuilder` locally it works ok but not via `pytest`. Any ideas?

mmcky commented:
thanks for your feedback @choldgraf I think this is now ready to go -- and testing picked up some good edge conditions 👍

choldgraf commented on jupyter_book/config.py:
```diff
@@ -286,18 +293,19 @@ def yaml_to_sphinx(yaml: dict):
                 sphinx_config["extensions"].append(extension)
 
     local_extensions = yaml.get("sphinx", {}).get("local_extensions")
+    add_paths = []
```
can you document what `add_paths` is in the docstring of this function so that other maintainers know what it's doing?

mmcky replied:
thanks @choldgraf this is done.

choldgraf replied:
Awesome! One quick question - if we just add to sys.path will it accomplish the same thing? (Eg without having to pass sys.path in and out of this function)?

mmcky replied:
this is what I had before the change -- and it worked for running `jupyter-book` locally but failed the `tests` as the path to the test collection wasn't correct in `sys.path`. It was adding the extension to a path at the root level of the test (rather than including the test set folder). I tracked this down to not resolving relative references when a `srcpath` is provided to jupyter-book. This update ensures the path is relative to `_config.yml` and uses `pathlib` to resolve the relative references that will often be specified in `_config.yml`

mmcky replied:
I needed to pass it up to the calling function as `yaml_path` was available in that context.

choldgraf commented:
ah great - thanks for clarifying this @mmcky - alright let's 🚢 it and see how this works for you 👍 hooray advanced features for power users!
